sharding/contracts/sharding_manager.sol

```diff
@@ -1,253 +1,326 @@
-    isCollatorDeposited[collators[_collatorIndex].addr] = false;
-    delete collators[_collatorIndex];
-    stackPush(_collatorIndex);
-    --numCollators;
-    Withdraw(_collatorIndex);
+  function compute_header_hash(uint256 shard_id,
+                               bytes32 parent_hash, 
+                               bytes32 chunk_root, 
+                               uint256 period,
+                               address proposer_address,
+                               uint256 proposer_bid) public returns(bytes32){
+
   }
 
-  // Attempts to process a collation header, returns true on success, reverts on failure.
-  function addHeader(int _shardId, uint _expectedPeriodNumber, bytes32 _periodStartPrevHash,
-                     bytes32 _parentHash, bytes32 _transactionRoot,
-                     address _coinbase, bytes32 _stateRoot, bytes32 _receiptRoot,
-                     int _number) public returns(bool) {
-    HeaderVars memory headerVars;
-
-    // Check if the header is valid
-    require((_shardId >= 0) && (_shardId < shardCount));
-    require(block.number >= periodLength);
-    require(_expectedPeriodNumber == block.number / periodLength);
-    require(_periodStartPrevHash == block.blockhash(_expectedPeriodNumber * periodLength - 1));
-
-    // Check if this header already exists
-    headerVars.entireHeaderHash = keccak256(_shardId, _expectedPeriodNumber, _periodStartPrevHash,
-                                   _parentHash, _transactionRoot, bytes32(_coinbase),
-                                   _stateRoot, _receiptRoot, _number);
-    assert(collationHeaders[_shardId][headerVars.entireHeaderHash].score == 0);
-    // Check whether the parent exists.
-    // if (parent_collation_hash == 0), i.e., is the genesis,
-    // then there is no need to check.
-    if (_parentHash != 0x0)
-        assert(collationHeaders[_shardId][_parentHash].score > 0);
-    // Check if only one collation in one period
-    assert(periodHead[_shardId] < int(_expectedPeriodNumber));
-
-    // Check the signature with validation_code_addr
-    headerVars.collatorAddr = getEligibleCollator(_shardId, block.number/periodLength);
-    require(headerVars.collatorAddr != 0x0);
-    require(msg.sender == headerVars.collatorAddr);
-
-    // Check score == collationNumber
-    headerVars.score = collationHeaders[_shardId][_parentHash].score + 1;
-    require(_number == headerVars.score);
-
-    // Add the header
-    collationHeaders[_shardId][headerVars.entireHeaderHash] = CollationHeader({
-      parentHash: _parentHash,
-      score: headerVars.score
-    });
+  function register_notary() public payable returns(bool) {
+    address notary_address = msg.sender;
+    require(!notary_registry[notary_address].deposited);
+    require(msg.value == NOTARY_DEPOSIT);
+
+    // update notary_sample_size
+    update_notary_sample_size();
+    
+    uint index;
+    if (!empty_stack()) {
```

Comment:
Can you invert this if statement so you don't have to negate the `empty_stack()`?

```
if (empty_stack()) {
  // do thing to empty stack
} else {
  // do thing to non-empty stack
}
```